### ISSUE TITLE
[OpenTK] DPI scaling is ignored by MonoGame

### ISSUE THREAD

**Issue by mrhelmut** [mono/MonoGame#4454]
Windows (with 4K displays) and OS X (with Retina displays) are using a high DPI mode which scales everything to fit high resolutions.

OpenTK has internal mechanisms to handle that automatically, but MonoGame overrides the ```Viewport``` size (because it assumes it to be the same as the backbuffer, which is not correct when it comes to DPI scaling) and this results in cropped or too big viewport.

I know that OpenTK is being replaced, but I feel like this is still an important issue and I'm leaving this here so we can keep that in mind when SDL2 will be in use (note that SDL2 <= 2.0.2 doesn't have DPI scaling support).

**Comment by cra0zy** [comment]
> I know that OpenTK is being replaced, but I feel like this is still an important issue and I'm leaving this here so we can keep that in mind when SDL2 will be in use (note that SDL2 <= 2.0.2 doesn't have DPI scaling support).

I am using a function from SDL 2.0.4 so... yea.. we will be using SDL 2.0.4.

**Comment by mrhelmut** [comment]
Awesome. :+1:

**Comment by tomspilman** [comment]
I suspect the Windows DX desktop build also has issues with high DPI displays.  We will have to find a device and do some testing.

**Comment by mrhelmut** [comment]
I'm suspecting that too, but it should be less complicated to fix on Windows: (from opentk/opentl#47 )

> Windows APIs are defined in terms of unscaled pixels, i.e. desktop resolution, window size, mouse position refer to pixels directly. A 2560x1600 monitor at 200% DPI scaling has a resolution of 2560x1600. A 1280x800 window would take up 50% of this monitor.
> Mac APIs (Cocoa), on the other hand, are defined in terms of scaled points: a 2560x1600 monitor at 200% DPI scaling has a reported resolution of 1280x800. A 1280x800 window would take up 100% of this monitor.
> Linux behaves in a similar fashion to Windows.
> SDL is even more complicated: 2.0.0 does not support scaling at all. 2.0.1 supports scaling on Mac OS but not Windows or Linux. Scaling for Window will probably be added in 2.0.2.

**Comment by tgjones** [comment]
There's a workaround on WindowsDX, which I mentioned here:
https://github.com/mono/MonoGame/issues/3358

**Comment by tomspilman** [comment]
> because adding an application manifest, with  dpiAware="true" , to my game project does fix the original mouse position bug. 

That workaround is great to know @tgjones !

**Comment by mrhelmut** [comment]
It looks like everything is correct in both MonoGame and OpenTK. Viewport and backbuffer have the good size.

What seems to be causing so much trouble is effects and transform matrix. When I draw a sprite with a scale matrix passed by ```SpriteBatch.Begin()``` and a custom pixel shader which does nothing but output the sprite as-is, the sprite gets scaled by the passed matrix AND the DPI scale.
If I don't pass any matrix to ```SpriteBatch.Begin()```, everything renders at the intended scale.

Is there any sprite batch / effect mechanism that could be affected by DPI? Maybe a difference between DX and GL which could cause this?

**Comment by tomspilman** [comment]
> the sprite gets scaled by the passed matrix AND the DPI scale.

What are you seeing happen?  Like if you draw a 10x10 pixel sprite what do you get?

> Is there any sprite batch / effect mechanism that could be affected by DPI? 

None that I can think of.  Still I suggest you step thru the code of a simple example to be sure i'm not forgetting anything.

> Maybe a difference between DX and GL which could cause this?

There is the OpenGL position fixup code:

https://github.com/mono/MonoGame/blob/develop/MonoGame.Framework/Graphics/GraphicsDevice.OpenGL.cs#L754

But that really shouldn't be doing anything more than a little pixel rounding.

**Comment by mrhelmut** [comment]
Looks like I like to see problems where there is none.

The issue is that the DPI scaling is done after the window is made visible, and only with the SDL2 backend. Hence, if you initialize size dependent objects before that, MonoGame returns the unscaled viewport/backbuffer, which was causing trouble to fullscreen effect render targets in my case.

This can be worked around by postponing size dependent initializations after ```Game.Initialize()``` and ```Game.LoadContent()```.

I'll try to check if I can patch that, but chances are that my efforts will be in favor to helping replacing opentk.

**Comment by mrhelmut** [comment]
After a lot of investigations, I'm concluding that the best route is to use the SDL2 backend on every platforms, leave OpenTK to manage the DPI and use a virtual resolution instead of trying to change the backbuffer / viewport.

The one and only remaining issue with OpenTK on OS X is that mouse coordinates are not scaled with the Retina scaling. I've submitted a PR to OpenTK to fix that MonoGame/openTK#3

Since this is not MonoGame related and that OpenTK is being replaced, I'm closing this issue and will help getting the new SDL2 backend to work correctly when it gets in.

Regarding Windows, making the process DPI-aware through the manifest does the job.

I also found out that SDL2 HiDPI support only works on OS X and iOS is not implemented on any other platforms as of 2.0.4 (the ALLOW_HIDPI flag does nothing outside of Apple platforms).
